labelbox/data/metrics/iou.py

```diff
@@ -1,281 +1,254 @@
-                                                 NDClassification, NDTool,
-                                                 NDMask, NDPoint, NDPolygon,
-                                                 NDPolyline, NDRadio, NDText,
-                                                 NDRectangle)
-from labelbox.data.metrics.preprocess import (create_schema_lookup,
-                                              url_to_numpy)
+from ..annotation_types import (Label, ObjectAnnotation,
+                                ClassificationAnnotation, Mask, Geometry, Point,
+                                Line, Checklist, Text, Radio)
 
-VectorTool = Union[NDPoint, NDRectangle, NDPolyline, NDPolygon]
-ClassificationTool = Union[NDText, NDRadio, NDChecklist]
 
-
-def mask_miou(predictions: List[NDMask], labels: List[NDMask]) -> float:
+def data_row_miou(ground_truth: Label, prediction: Label) -> Optional[float]:
     """
-    Creates prediction and label binary mask for all features with the same feature schema id.
+    Calculate iou for two labels corresponding to the same data row.
 
     Args:
-        predictions: List of masks objects
-        labels: List of masks objects
+        ground_truth : Label containing human annotations or annotations known to be correct
+        prediction: Label representing model predictions
     Returns:
-        float indicating iou score
+        float indicating the iou score for this data row.
+        Returns None if there are no annotations in ground_truth or prediction Labels
     """
+    return get_iou_across_features(ground_truth.annotations,
+                                   prediction.annotations)
 
-    colors_pred = {tuple(pred.mask['colorRGB']) for pred in predictions}
-    colors_label = {tuple(label.mask['colorRGB']) for label in labels}
-    error_msg = "segmentation {} should all have the same color. Found {}"
-    if len(colors_pred) > 1:
-        raise ValueError(error_msg.format("predictions", colors_pred))
-    elif len(colors_label) > 1:
-        raise ValueError(error_msg.format("labels", colors_label))
-
-    pred_mask = _instance_urls_to_binary_mask(
-        [pred.mask['instanceURI'] for pred in predictions], colors_pred.pop())
-    label_mask = _instance_urls_to_binary_mask(
-        [label.mask['instanceURI'] for label in labels], colors_label.pop())
-    assert label_mask.shape == pred_mask.shape
-    return _mask_iou(label_mask, pred_mask)
 
-
-def classification_miou(predictions: List[ClassificationTool],
-                        labels: List[ClassificationTool]) -> float:
+def get_iou_across_features(
+    ground_truths: List[Union[ObjectAnnotation, ClassificationAnnotation]],
+    predictions: List[Union[ObjectAnnotation, ClassificationAnnotation]]
+) -> Optional[float]:
     """
-    Computes iou for classification features.
+    Groups annotations by schema_id or name (which is available), calculates iou score and returns the mean across all features.
```

Comment:
Would it be useful to return a mapping of features to per-feature iou score?

Reply:
I think in the future we can create a function for that and basically move most of this logic to that function. For now this make sense since the aggregation is used for both top level and subclasses.